### PR TITLE
add DiceLoss

### PR DESCRIPTION
**Pull request by tcapelle**
Add DiceLoss for segmentation tasks.
It is very useful to combine this type of loss with CE or FocalLoss, used in various kaggle comps.

### CODE REVIEW

File: fastai/losses.py [hunk 1]
```diff
@@ -122,4 +122,24 @@ class LabelSmoothingCrossEntropyFlat(BaseLoss):
...
+    "Target mask to one hot"
+    return torch.stack([torch.where(x==c, 1,0) for c in range(classes)], axis=axis)
+
+# Cell
+class DiceLoss:
+    "Dice loss for segmentation"
+    def __init__(self, axis=1, smooth=1):
+        store_attr()
+    def __call__(self, pred, targ):
+        targ = _one_hot(targ, pred.shape[self.axis])
```
Comment (jph00) [inline fastai/losses.py:138]: Why not use the `one_hot` from `torch_core`?

File: nbs/01a_losses.ipynb [hunk 2]
```diff
@@ -414,6 +414,113 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "dl = DiceLoss()\n",
+    "output = torch.randn(32, 4, 5, 10)\n",
+    "target = torch.randint(0,2,(32, 5, 10))\n",
+    "_ = dl(output, target)"
```
Comment (jph00) [inline nbs/01a_losses.ipynb:467]: Could you add some simple example tests that confirm the loss is correct? Currently your tests aren't actually testing anything.

**Review by jph00 — CHANGES_REQUESTED** [review]

File: fastai/losses.py (hunk 1 above)
Comment (tcapelle) [reply fastai/losses.py:138]: I know we have one, but is does not work the same. I would need to modify the other one to be able to create the one-hot segmentation masks.

File: nbs/01a_losses.ipynb (hunk 2 above)
Comment (tcapelle) [reply nbs/01a_losses.ipynb]: Would you mind helping me with this? I added some test for the `one_hot` and that the decodes work, but how would you unit test this? failing mostly happen with `flatten_check` so I would just be testing that.

File: fastai/losses.py (hunk 1 above)
Comment (tcapelle) [reply fastai/losses.py:138]: the `DiceMulti` metric already does this:
```python
        for c in range(learn.pred.shape[self.axis]):
            p = torch.where(pred == c, 1, 0)
```

File: nbs/01a_losses.ipynb (hunk 2 above)
Comment (jph00) [reply nbs/01a_losses.ipynb]: I'd test stuff like two masks that were identical, two which were entirely non-overlapping, and two which overlapped 50% exactly.

File: fastai/losses.py (hunk 1 above)
Comment (jph00) [reply fastai/losses.py:138]: What's the difference needed in functionality exactly? Is there some different name you could use for yours that makes it clear? And/or some comment you could add?

Comment (tcapelle) [reply fastai/losses.py:138]: I could make it a class method, I just put it outside to test it =P

File: nbs/01a_losses.ipynb (hunk 2 above)
Comment (tcapelle) [reply nbs/01a_losses.ipynb]: good idea

Comment (tcapelle) [reply nbs/01a_losses.ipynb]: done.

**Comment by tcapelle** [comment]
It is ready to review @jph00 (I initially solved the CI that was not passing by cleaning tutorial.text, but maybe was not a good idea).

**Comment by jph00** [comment]
Many thanks! :)
